docs/lang/logic.md

```diff
@@ -0,0 +1,458 @@
+                                    \* type error in Apalache
+  { { 1, 2 } } = { { 1, 2, 2, 2 } } \* TRUE
+  <<1, "a">> = <<1, "a">>           \* TRUE
+  <<1, "a">> = <<1, "b">>           \* FALSE
+  <<1, FALSE>> = <<2>>              \* FALSE in pure TLA+ and TLC,
+                                    \* type error in Apalache
+  <<1, 2>> = <<1, 2, 3>>            \* FALSE in pure TLA+ and TLC,
+                                    \* FALSE in Apalache, when both values
+                                    \* are treated as sequences
+  [ a |-> 1, b |-> 3 ] = [ a |-> 1, b |-> 3 ]           \* TRUE
+  [ a |-> 1, b |-> 3 ] = [ a |-> 1 ]                    \* FALSE
+  [ x \in 2..2 |-> x + x ] = [ x \in {2} |-> 2 * x ]    \* TRUE
+  [ x \in 2..3 |-> x + x ] = [ x \in {2, 3} |-> 2 * x ] \* FALSE
+```
+
+**Example in Python:** The standard data structures also implement
+    structural equality in Python, though we have to be careful to
+    use `==` instead of `=`:
+
+```python
+  False == False
+  False == True
+  10 == 20
+  15 == 15
+  "Hello" == "world"
+  "Hello" == "hello"
+  "Bob" == "Bob"
+  frozenset({ 1, 2 }) == frozenset({ 2, 3 })
+  frozenset({ 1, 2 }) == frozenset({ 2, 1 })
+  frozenset({ 1 }) - frozenset({ 1 }) == frozenset({ "a" }) - frozenset({ "a" })
+  frozenset({ frozenset({ 1, 2 }) }) == frozenset({ frozenset({ 1, 2, 2, 2 }) })
+  (1, "a") == (1, "a")
+  (1, "a") == (1, "b")
+  (1, False) == (2, )
+  (1, 2) == (1, 2, 3)
+  { "a": 1, "b": 3 } == { "a": 1, "b": 3 }
+  { "a": 1, "b": 3 } == { "a": 1 }
+  { x: (x + x) for x in { 2 } } == { x: (x * x) for x in { 2 } }
+  { x: (x + x) for x in { 2, 3 } } == { x: 2 * x for x in { 2, 3 } } 
+```
+
+----------------------------------------------------------------------------
+
+<a name="neq"></a>
+### Inequality
+
+**Notation:** `e_1 /= e_2` or `e_1 # e_2`
+
+**LaTeX notation:** ![ne](./img/ne.png)
+
+**Arguments:** Two arguments.
+
+**Effect:** This operator is syntax sugar for `~(e_1 = e_2)`. Full stop.
+
+----------------------------------------------------------------------------
+
+<a name="chooseBounded"></a>
+### Bounded Choice
+
+_This operator causes a lot of confusion. Read carefully!_
+
+**Notation:** `CHOOSE x \in S: P`
+
+**LaTeX notation:** ![choose-bounded](./img/choose-bounded.png)
+
+**Arguments:** Three arguments: a variable name, a set, and an
+expression.
+
+**Effect:** This operator is a magical device by the definition, no kidding. It
+is a blackbox algorithm that _somehow_ picks one element `e` of `S` that
+satisfies the expression `P` against the binding `[x |-> e]` and returns this
+element `e`.
+
+Did we say "an algorithm"? Yes! `CHOOSE x \in S: P` is deterministic.  If you
+give it two equal sets and two equivalent predicates, it will give you the
+same value. More formally, the only known property of `CHOOSE` is as follows:
+
+```tla
+  {x \in S: P} = {y \in T: Q} =>
+      (CHOOSE x \in S: P) = (CHOOSE y \in T: Q)
+```
```

Comment:
Is this really true? Take `S == {1}` and `T == {1,2}`, then `P == TRUE` and `Q == y = 1`. Obviously `S /= T` and `~(P <=> Q) `, even though `{x\ in {1}: TRUE} = {y \in {1,2}: y = 1}`. 
My understanding is that for `P <=> Q` it holds that `(CHOOSE x \in S: P) = (CHOOSE x \in S: Q)` and for `S = T`, then `(CHOOSE x \in S: P) = (CHOOSE x \in T: P)`, but I'm not sure you can combine those two properties.

Reply:
The book defines `CHOOSE x \in S` as `CHOOSE (x \in S) /\ P`, so it should be true. I am not sure about TLC though. Of course, what is written in English is slightly more restrictive.